src/main/java/io/github/hotlava03/chatutils/listeners/AntiSpamListener.java

```diff
@@ -1,34 +1,102 @@
+import net.minecraft.util.math.MathHelper;
 
 import java.util.function.Consumer;
 
-public class AntiSpamListener implements Consumer<ReceiveMessageEvent> {
-    private Text lastMessage;
-    private int spamCounter = 1;
+import static io.github.hotlava03.chatutils.util.OrderedTextAdapter.orderedTextToString;
+import static io.github.hotlava03.chatutils.util.StringUtils.getDifference;
+import static io.github.hotlava03.chatutils.util.StringUtils.isNumeric;
 
+public class AntiSpamListener implements Consumer<ReceiveMessageEvent> {
     @Override
     public void accept(ReceiveMessageEvent e) {
         if (ChatStorage.getInstance().isBlockingChatEvents()) return;
-        if (ChatUtilsConfig.ANTI_SPAM.value()) {
-            if (this.lastMessage == null) {
-                this.lastMessage = e.getText();
-                this.spamCounter = 1;
-                return;
+        var client = MinecraftClient.getInstance();
+        var chat = client.inGameHud.getChatHud();
+        var serverInfo = client.getCurrentServerEntry();
+        var address = serverInfo != null ? serverInfo.address : null;
+        var fullHistory = e.getLines();
+        var range = ChatUtilsConfig.ANTI_SPAM_RANGE.value();
+        var history = fullHistory.size() >= range ? fullHistory.subList(0, range) : fullHistory;
+        if (history.isEmpty()) return;
+
+        int maxTextLength = MathHelper.floor(chat.getWidth() / chat.getChatScale());
+        var splitLines = ChatMessages.breakRenderedChatMessageLines(
+                e.getText(), maxTextLength, client.textRenderer);
+
+        int spamCounter = 1;
+        int lineMatchCount = 0;
+
+        for (int i = history.size() - 1; i >= 0; i--) {
+            var previous = orderedTextToString(history.get(i).content());
+
+            if (lineMatchCount <= splitLines.size() - 1) {
+                String next = orderedTextToString(splitLines.get(lineMatchCount));
+
+                if (lineMatchCount < splitLines.size() - 1) {
+                    if (getDifference(previous, next) <= 0) lineMatchCount++;
+                    else lineMatchCount = 0;
+
+                    continue;
+                }
+
+                if (!previous.startsWith(next)) {
+                    lineMatchCount = 0;
+                    continue;
+                }
+
+                if (i > 0 && lineMatchCount == splitLines.size() - 1) {
+                    var appended = (previous + orderedTextToString(history.get(i - 1).content()))
+                            .substring(next.length());
+
+                    if (appended.startsWith(" [x") && appended.endsWith("]")) {
+                        var previousCounter = appended.substring(3, appended.length() - 1);
+
+                        if (isNumeric(previousCounter)) {
+                            spamCounter += Integer.parseInt(previousCounter);
+                            lineMatchCount++;
+                            continue;
+                        }
```

Comment:
you know you could save some CPU cycles and avoid writing a few extra lines of code just by doing
```suggestion
                        try {
                            spamCounter += Integer.parseInt(previousCounter);
                            lineMatchCount++;
                            continue;
                        } catch (NumberFormatException ignored) {}
```
now i have a reason to dislike `isNumeric` functions just because they're misused most of the time